### PR TITLE
add scripts to generate pdf

### PR DESCRIPTION
**Pull request by LeoMcA**
### Description

Add scripts to generate a PDF of the curriculum using Pandoc and some custom LaTeX commands/Lua scripting.

Here's the PDF generated on the current content: [MDN-Curriculum-2024-09.pdf](https://github.com/user-attachments/files/17043814/MDN-Curriculum-2024-09.pdf)

### Motivation

Necessary for https://github.com/mdn/curriculum/pull/71

### Additional details

https://mozilla-hub.atlassian.net/browse/MP-1483

### Related issues and pull requests

- https://github.com/mdn/curriculum/pull/71

### CODE REVIEW

**Comment by argl** [comment]
Some non-technical comments:

The PDF looks much better now - typography FTW!

Not a technical comment, but some things that went through my mind when looking at the document:

It looks super unbranded, I am missing something like the MDN logo on the front page, and probably a prominent link back to the MDN/curriculum site. Since this is probably handed around on private channels, something to remind people where this came from would help. Also, only a single mention of "Mozilla" in the whole thing, which is kind of sad.

Generally, the front page could use some design love, maybe not centering the title (centering is for legal stuff, certificates, scientific papers and death notices, I have learned). Justifying the normal text blocks to full width is debatable, but with these book-like long lines it is not too bad. 

I have seen a "Notes" box having a page break right after the "Notes" header (page 14/15) which is something a human typesetter from the last century would have avoided by putting the whole box to the next page. The rule in typesetting software is usually along the lines of "if there is a page break right after a heading, move the heading to the next page so that it stays with its content".

**Review by chrisdavidmills — COMMENTED** [review]
@LeoMcA Couple of notes for you on the PDF design — these are not major showstoppers, and I'm not sure how easy they would be to fix. Maybe as an iterative process?

- The note boxes feel like they could do with a bit more padding.
- In general, the paragraphs and bullets could do with a bit more line-height. The recommendation is usually 1.5.
- In the TOC and the document, the top-level headings (e.g. "Getting started modules") could do with having a larger font size than the ones below (e.g. "1. Soft skills")
- Third-level sub-bullets (e.g. on Pages 5, 25, 35, 41, 47, 48) use standard asterisk characters, which look pretty crappy. There are not many of these, fortunately.
- Further to what Andi said, there is also a note box broken pretty badly just after the heading on page 34/35.

Aside: On looking through the PDF, I noticed that mostly we don't link to many code terms (e.g. `length`, `indexOf()`), although weirdly we do link to a couple (e.g. `fetch()`). I am thinking that maybe we should, as a convenience fo readers. This would be quite a lot of work for me to update in the curriculum source, but it would make for a more usable set of online docs and PDFs. What do folks think?

Also, you asked me to think about what wording would be good for the last updated date line, considering we have small updates (e.g. typos), but also major revisions, i.e. the review we intend to do each June-ish.

I would suggest having the third line in the PDF read like this:

> PDF generated: September 18, 2024 | Last major curriculum review: February 2024

I'm assuming the first date could be auto-generated, but we'd need to manually update the second date?

**Comment by LeoMcA** [comment]
Thanks for the notes, will work on another iteration of the PDF before launch.

I've added a github action to automatically build and release the PDF on each push to main, here's how it looks: https://github.com/LeoMcA/curriculum/releases

File: pdf/build.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+
+cd "$(dirname "$0")"
+
+pandoc header.md ../curriculum/**/*.md -s \
+  -o MDN-Curriculum.pdf \
+  --metadata date="`date -u '+%Y-%m-%d'`" \
+  --pdf-engine=xelatex \
+  --lua-filter=adjust-links.lua \
+  --include-in-header=header.tex
```
Comment (caugner) [inline pdf/build.sh:10]:
```suggestion
  --include-in-header=header.tex

```

File: pdf/build.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
```
Comment (caugner) [inline pdf/build.sh:1]:
```suggestion
#!/usr/bin/env sh
```

File: pdf/header.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+---
+title: MDN Curriculum
```
Comment (caugner) [inline pdf/header.md:2]:
```suggestion
# See: https://pandoc.org/MANUAL.html#variables
title: MDN Curriculum
```

File: pdf/adjust-links.lua [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      id = "#" .. id
+    end
+  end
+
+  -- Fix section header links
+  if id:sub(1,1) == '#' then
+    id = id:gsub("_", "-")
+    id = id:gsub("^[^a-zA-Z]+", "")
+    id = "#" .. id
+  end
```
Comment (caugner) [inline pdf/adjust-links.lua:19]: Can you before/after examples here?

**Review by caugner — APPROVED** [review]
LGTM, just some nits.

**Comment by LeoMcA** [comment]
Should've addressed all comments on the PDF other than:

> In the TOC and the document, the top-level headings (e.g. "Getting started modules") could do with having a larger font size than the ones below (e.g. "1. Soft skills")

Due to how our content is structured, this isn't possibly very easily, so I've left it how it is.

I'll merge now, which should kick off a build of the PDF as a github release: we can address any further PDF issues with a new PR.

**Pull request merged**